File: modules/version-consistency-checker/src/version_consistency_checker/checker.py
from pydantic import BaseModel

from version_consistency_checker.utils.normalize_tool_name import normalize_tool_name


class Mismatch(BaseModel):
  tool: str
  nix_version: str
  mismatched_sources: dict[str, str]  # source name -> version


def check_consistency(
  nix_versions: dict[str, str],
  **sources: dict[str, str],
) -> list[Mismatch]:
  mismatches: list[Mismatch] = []

  for tool, nix_version in nix_versions.items():
    mismatched: dict[str, str] = {}

    for source_name, raw_versions in sources.items():
      normalized = {normalize_tool_name(k): v for k, v in raw_versions.items()}
      source_version = normalized.get(tool)

      if source_version is None:
        continue

      if source_version != nix_version:
        mismatched[source_name] = source_version

    if mismatched:
      mismatches.append(
        Mismatch(tool=tool, nix_version=nix_version, mismatched_sources=mismatched)
      )

  return mismatches

File: modules/version-consistency-checker/src/checker.ts
import { normalizeToolName } from "./utils/normalizeToolName.ts";

export interface Mismatch {
  tool: string;
  nixVersion: string;
  mismatchedSources: Record<string, string>;
}

/**
 * Checks version consistency between nix devshell and one or more sources.
 *
 * For each tool in nixVersions, compares against each source after normalizing
 * source tool names. Returns mismatches where at least one source disagrees.
 */
export const checkConsistency = (
  nixVersions: Record<string, string>,
  sources: Record<string, Record<string, string>>,
): Mismatch[] => {
  const mismatches: Mismatch[] = [];

  for (const [tool, nixVersion] of Object.entries(nixVersions)) {
    const mismatchedSources: Record<string, string> = {};

    for (const [sourceName, rawVersions] of Object.entries(sources)) {
      const normalized = Object.fromEntries(
        Object.entries(rawVersions).map(([k, v]) => [normalizeToolName(k), v]),
      );
      const sourceVersion = normalized[tool];

      if (sourceVersion === undefined) {
        continue;
      }

      if (sourceVersion !== nixVersion) {
        mismatchedSources[sourceName] = sourceVersion;
      }
    }

    if (Object.keys(mismatchedSources).length > 0) {
      mismatches.push({ tool, nixVersion, mismatchedSources });
    }
  }

  return mismatches;
};
